src/control/jobs/XournalScheduler.cpp

```diff
@@ -12,51 +12,60 @@ XournalScheduler::XournalScheduler()
 XournalScheduler::~XournalScheduler() = default;
 
 void XournalScheduler::removeSidebar(SidebarPreviewBaseEntry* preview) {
-    // Dont await for tasks, as it completly blocks whole program.
-    // Not sure why.
-    removeSource(preview, JOB_TYPE_PREVIEW, JOB_PRIORITY_HIGH, false);
+    removeSource(preview, JOB_TYPE_PREVIEW, JOB_PRIORITY_HIGH, true);
 }
 
 void XournalScheduler::removePage(XojPageView* view) { removeSource(view, JOB_TYPE_RENDER, JOB_PRIORITY_URGENT); }
 
 void XournalScheduler::removeAllJobs() {
-    g_mutex_lock(&this->jobQueueMutex);
+    std::lock_guard<std::mutex> lock(this->jobQueueMutex);
 
     for (int priority = JOB_PRIORITY_URGENT; priority < JOB_N_PRIORITIES; priority++) {
-        int length = g_queue_get_length(this->jobQueue[priority]);
-        for (int i = 0; i < length; i++) {
-            Job* job = static_cast<Job*>(g_queue_peek_nth(this->jobQueue[priority], i));
+        std::deque<Job*>& queue = *this->jobQueue[priority];
+        auto it = queue.begin();
 
+        while (it != queue.end()) {
+            Job* job = *it;
+
+            // Only remove PREVIEW and RENDER jobs; we aren't
+            // responsible for other types of jobs.
             JobType type = job->getType();
             if (type == JOB_TYPE_PREVIEW || type == JOB_TYPE_RENDER) {
                 job->deleteJob();
-                g_queue_remove(this->jobQueue[priority], job);
+
+                it = queue.erase(it);
+
                 job->unref();
+                job = nullptr;
+            } else {
+                ++it;
             }
         }
     }
-
-    g_mutex_unlock(&this->jobQueueMutex);
 }
 
 void XournalScheduler::finishTask() {
-    g_mutex_lock(&this->jobRunningMutex);
-    g_mutex_unlock(&this->jobRunningMutex);
+    this->jobRunningMutex.lock();
+    this->jobRunningMutex.unlock();
```

Comment:
We may want to use `std::condition_variable::wait`

Reply:
also, don't call lock/unlock manually
```suggestion
    std::unique_lock guard g{this->jobRunningMutex};
```

Reply:
I think locking on the `jobRunningMutex` is better here; we don't want this to block when a job isn't already running.

I made the suggested change, though!